File: _deploy_vercel_sync/sovereign-core/core/scout.py
"""
SOVEREIGN V14 - Scout Module
Google News RSS Tarayıcı (Haber Bulucu)
"""

import time
import hashlib
import logging
from typing import List
from datetime import datetime

import feedparser
from pydantic import BaseModel

logger = logging.getLogger(__name__)


class NewsItem(BaseModel):
    """Haber öğesi veri modeli"""
    id: str
    title: str
    link: str
    published: str
    content: str
    source: str
    hash: str


class Scout:
    """RSS Haber Tarayıcı"""
    
    def __init__(self, keywords: List[str]):
        self.keywords = keywords
        logger.info(f"[SCOUT] {len(keywords)} anahtar kelime ile başlatıldı")
    
    def generate_rss_feeds(self) -> List[str]:
        """Google News RSS feed URL'lerini oluştur"""
        feeds = []
        for keyword in self.keywords:
            encoded_keyword = keyword.replace(' ', '+')
            url = f"https://news.google.com/rss/search?q={encoded_keyword}&hl=en-US&gl=US&ceid=US:en"
            feeds.append(url)
        return feeds
    
    def generate_hash(self, title: str, link: str) -> str:
        """Deduplication için MD5 hash oluştur"""
        return hashlib.md5(f"{title}{link}".encode()).hexdigest()
    
    def fetch_feed(self, feed_url: str) -> List[NewsItem]:
        """RSS feed'i çek ve parse et"""
        try:
            logger.info(f"[SCOUT] Taranıyor: {feed_url[:80]}...")
            feed = feedparser.parse(feed_url)
            
            news_items = []
            for entry in feed.entries[:5]:  # Feed başına 5 haber
                news_item = NewsItem(
                    id=f"news-{int(time.time())}-{len(news_items)}",
                    title=entry.get('title', 'Başlık yok'),
                    link=entry.get('link', ''),
                    published=entry.get('published', datetime.now().isoformat()),
                    content=entry.get('summary', entry.get('description', '')),
                    source=entry.get('source', {}).get('title', 'Bilinmeyen'),
                    hash=self.generate_hash(entry.get('title', ''), entry.get('link', ''))
                )
                news_items.append(news_item)
            
            logger.info(f"[SCOUT] {len(news_items)} haber bulundu")
            return news_items
            
        except Exception as e:
            logger.error(f"[SCOUT] Feed hatası: {e}")
            return []
    
    def aggregate_news(self) -> List[NewsItem]:
        """Tüm RSS feed'lerden haberleri topla"""
        logger.info("[SCOUT] Haber toplama başladı...")
        
        feeds = self.generate_rss_feeds()
        all_news = []
        
        for feed_url in feeds:
            items = self.fetch_feed(feed_url)
            all_news.extend(items)
            time.sleep(1)  # Feed'ler arası rate limit
        
        logger.info(f"[SCOUT] Toplam {len(all_news)} haber toplandı")
        return all_news

File: lib/sovereign-core/scout.ts
/**
 * SCOUT - RSS Aggregator
 * Google News / RSS feed'lerden finansal istihbaratı toplar.
 * Gerçek RSS parse: rss-parser. Başarısız olursa mock fallback.
 */

import crypto from 'crypto';
import Parser from 'rss-parser';

export interface NewsItem {
  id: string;
  title: string;
  link: string;
  pubDate: string;
  content: string;
  source: string;
  hash: string;
}

const KEYWORDS = [
  'Nasdaq',
  'Bitcoin',
  'Fed',
  'AI',
  'Crypto Regulation',
  'Federal Reserve',
  'Cryptocurrency',
  'Stock Market',
  'Artificial Intelligence'
];

/**
 * Google News RSS feed URL'lerini oluşturur
 */
function generateRSSFeeds(): string[] {
  return KEYWORDS.map(keyword =>
    `https://news.google.com/rss/search?q=${encodeURIComponent(keyword)}&hl=en-US&gl=US&ceid=US:en`
  );
}

/**
 * Haber içeriğinden hash oluşturur (deduplication için)
 */
export function generateNewsHash(title: string, link: string): string {
  return crypto
    .createHash('md5')
    .update(`${title}${link}`)
    .digest('hex');
}

const parser = new Parser({
  timeout: 8000,
  headers: {
    'User-Agent': 'Mozilla/5.0 (compatible; SIA-Scout/1.0; +https://siaintel.com)'
  }
});

/**
 * RSS feed'i gerçek parse eder; hata olursa mock döner
 */
export async function fetchRSSFeed(feedUrl: string): Promise<NewsItem[]> {
  try {
    const feed = await parser.parseURL(feedUrl);
    if (!feed?.items?.length) return [];

    const items: NewsItem[] = feed.items.map((item, i) => {
      const title = item.title?.trim() || 'Untitled';
      const link = item.link?.trim() || item.guid || '';
      const content = item.contentSnippet?.trim() || item.content?.replace(/<[^>]+>/g, '').slice(0, 500) || title;
      const pubDate = item.pubDate ? new Date(item.pubDate).toISOString() : new Date().toISOString();
      const source = item.creator || (item as any).source?.name || 'RSS';
      return {
        id: `rss-${Date.now()}-${i}-${generateNewsHash(title, link).slice(0, 8)}`,
        title,
        link,
        pubDate,
        content,
        source,
        hash: generateNewsHash(title, link)
      };
    });

    console.log(`[SCOUT] Fetched ${items.length} items from ${feedUrl.slice(0, 60)}...`);
    return items;
  } catch (error) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn('[SCOUT] RSS fetch failed, using mock:', (error as Error).message);
      return getMockNewsItems();
    }
    console.warn('[SCOUT] RSS fetch failed:', (error as Error).message);
    return [];
  }
}

function getMockNewsItems(): NewsItem[] {
  const now = new Date().toISOString();
  const mock: NewsItem[] = [
    { id: `news-${Date.now()}-1`, title: 'Federal Reserve Signals Rate Cuts as Inflation Cools to 2.3%', link: 'https://example.com/fed-rate-cuts', pubDate: now, content: 'The Federal Reserve indicated potential interest rate cuts in Q2 2026 as core PCE inflation declined to 2.3%. Fed Chair Powell emphasized a data-dependent approach.', source: 'Reuters', hash: '' },
    { id: `news-${Date.now()}-2`, title: 'Bitcoin Surges Past $52,000 on Institutional Demand', link: 'https://example.com/bitcoin-surge', pubDate: now, content: 'Bitcoin rallied above $52,000 as institutional investors increased exposure. Trading volume reached $45B in 24 hours.', source: 'Bloomberg', hash: '' },
    { id: `news-${Date.now()}-3`, title: 'Nasdaq Rallies 3.2% on Tech Sector Optimism', link: 'https://example.com/nasdaq-rally', pubDate: now, content: 'The Nasdaq Composite surged 3.2% as tech stocks rallied on dovish Fed signals. Nvidia, Microsoft, and Apple led gains.', source: 'CNBC', hash: '' }
  ];
  return mock.map(item => ({ ...item, hash: generateNewsHash(item.title, item.link) }));
}

/**
 * Tüm RSS feed'lerden haberleri toplar
 */
export async function aggregateNews(): Promise<NewsItem[]> {
  console.log('[SCOUT] Starting news aggregation...');
  
  const feeds = generateRSSFeeds();
  const allNews: NewsItem[] = [];

  for (const feedUrl of feeds) {
    const items = await fetchRSSFeed(feedUrl);
    allNews.push(...items);
    
    // Rate limiting between feeds
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  console.log(`[SCOUT] Aggregated ${allNews.length} news items`);
  
  return allNews;
}

/**
 * Scout'u başlatır (her 15 dakikada bir çalışır)
 */
export function startScout(callback: (news: NewsItem[]) => void): NodeJS.Timeout {
  console.log('[SCOUT] Starting autonomous news scout...');
  
  // İlk çalıştırma
  aggregateNews().then(callback);
  
  // Her 15 dakikada bir çalıştır
  const interval = setInterval(async () => {
    const news = await aggregateNews();
    callback(news);
  }, 15 * 60 * 1000); // 15 minutes

  return interval;
}
